source/common/network/dns_impl.cc

```diff
@@ -173,7 +173,8 @@ void DnsResolverImpl::PendingResolution::onAresGetAddrInfoCallback(int status, i
       //  We can't add a main thread assertion here because both this code is reused by dns filter
       //  and executed in both main thread and worker thread. Maybe split the code for filter and
       //  main thread.
-      TRY_NEEDS_AUDIT { callback_(resolution_status, std::move(address_list)); }
+      TRY_ASSERT_MAIN_THREAD { callback_(resolution_status, std::move(address_list)); }
```

Comment:
Having 2 copies of the code that handles DNS resolutions just because of this try/catch issue doesn't seem good.

The use of dispatcher_ post to rethrow the exception in a different thread seems even stranger.  Who would catch the posted exception?

Can we require that the callback passed in to dns_impl never throw?

Reply:
+1 I think it's a non-starter to copy this code. Either we need to share all of it, or we should just change the semantics of exception handling within the callback per @antoniovicente (though this may be difficult).

> rethrow the exception in a different thread seems even stranger.

It's posting to the same thread. It's an artifact of not throwing through the C code. I agree, it's not good.

Reply:
Thanks, this is only a draft, I am not satisfied with the code duplication myself. l agree it is better to handle it without copying the code.

Reply:
Do we know which callback throws?

Reply:
The reason why we post to the same thread is because of #4307, c_ares as a c library has exception unsafe code, so exception throwing need to be executed in another call stack in the same thread.

Reply:
The callbeck can be in [logical_dns_cluster](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/common/upstream/logical_dns_cluster.cc#L104) or [strict_dns_cluster](https://github.com/envoyproxy/envoy/blob/b4b82411487d66290da72948b3090ad2ed605c96/source/common/upstream/strict_dns_cluster.cc#L111)
Network::Utility::getAddressWithPort may throw.  I haven't found any other exception throwing code yet.